### PR TITLE
JaversSnapshotsCompiler added to core functionality

### PR DESCRIPTION
**Pull request by smirnp**
For more details see https://github.com/javers/javers/issues/133

### CODE REVIEW

File: javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        //List<CdoSnapshot> snapshots = javers.getStateHistory(convertInstanceIdToDTO(instanceId), -1); // load all
+        //List<CdoSnapshot> snapshots = javers.findSnapshots(QueryBuilder.byInstanceId(instanceId, instanceId.getgetBaseJavaClass() ).build());
+        List<CdoSnapshot> snapshots = javers.findSnapshots(QueryBuilder.byInstanceId(instanceId, instanceId.getClass()).build());
+
+        double commitToBeSearched = commitIdAsDouble(commitId);
+        CdoSnapshot snapFound = null;
+        // since all snapshot are sorted from the highest commit value to the
+        // lowest one
+        for (CdoSnapshot snap : snapshots) {
+            if (commitIdAsDouble(snap.getCommitId()) < (commitToBeSearched + EPS)) {
```
Comment (bartoszwalacik) [inline javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java]: commitId has well defined, decimal value, why to play with doubles and what is EPS?

File: javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    private Object compileEntityForCommitId(InstanceId instanceId, CommitId commitId) throws IllegalAccessException,
+            InstantiationException {
+
+        // TODO here the Javers function for loading the CdoSnapshot for the
+        // given commitId should be called, but the current Javers
+        // implementation doesn't provide it
+        // TODO assumes that limit = -1 means "all CdoSnapshots"
+        //List<CdoSnapshot> snapshots = javers.getStateHistory(convertInstanceIdToDTO(instanceId), -1); // load all
+        //List<CdoSnapshot> snapshots = javers.findSnapshots(QueryBuilder.byInstanceId(instanceId, instanceId.getgetBaseJavaClass() ).build());
+        List<CdoSnapshot> snapshots = javers.findSnapshots(QueryBuilder.byInstanceId(instanceId, instanceId.getClass()).build());
```
Comment (bartoszwalacik) [inline javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java]: fetching all snapshots could kill performance, we can think about adding queryByCommitId to JQL...

File: javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    /**
+     * Loads the latest commited entity state for given instanceId
+     *
+     * @param instanceId
+     * @return
+     */
+    public Object compileLatestEntityState(InstanceId instanceId) {
+        InstanceIdDTO dto = convertInstanceIdToDTO(instanceId);
+        Optional<CdoSnapshot> latestSnapshot = javers.getLatestSnapshot(dto, dto.getClass());
+        //Optional<CdoSnapshot> latestSnapshot = javers.getLatestSnapshot()getLatestSnapshot(convertInstanceIdToDTO(instanceId));
```
Comment (bartoszwalacik) [inline javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java]: all commented code should be removed

File: javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+     * @param commitId
+     * @return
+     */
+    public Object compileEntityStateForCommitId(InstanceId instanceId,
+                                                String commitId) {
+        clearCache();
+        try {
+            return compileEntityForCommitId(instanceId, CommitId.valueOf(commitId));
+        } catch (IllegalAccessException | InstantiationException e) {
+            e.printStackTrace();
```
Comment (bartoszwalacik) [inline javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java]: public method in javers never returns null, just throw JaversRuntimeException

File: javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                        }
+                        for (Object element : toBeDeleted) {
+                            collection.remove(element);
+                        }
+                    } else if (propertyValue instanceof InstanceId) {
+                        propertyValueToSet = compileEntityForCommitId(
+                                (InstanceId) propertyValue, snap.getCommitId());
+                    }
+                    setter.invoke(instance, propertyValueToSet);
+                } catch (Exception exc) {
```
Comment (bartoszwalacik) [inline javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java]: swallowing all Exceptions is generally an antipatten, wrap it in JaversRuntimeException and rethrow

File: javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+import org.javers.repository.jql.InstanceIdDTO;
+import org.javers.repository.jql.QueryBuilder;
+
+/**
+ * Loads under certain assumptions the original entity state for given
+ * CdoSnapshot or CommitId.<br>
+ * 1) Loads the whole object graph eagerly (lazy loading would require the use
+ * of proxy objects). <br>
+ * 2) All entities have constructor with no arguments -> Java Beans standard. <br>
+ * 3) Each entity field ({@link JaversField}) has a setter method with the
```
Comment (bartoszwalacik) [inline javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java]: we should support both FIELD and BEAN property styles

File: javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
+package org.javers;
```
Comment (bartoszwalacik) [inline javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java]: there is no tests here, see how we write tests in Javers and add some
https://github.com/javers/javers/blob/master/javers-core/src/test/groovy/org/javers/core/JaversDiffE2ETest.groovy

**Comment by bartoszwalacik** [comment]
@smirnp i did a code review, please follow my suggestions and then i could merge your solution to javers-incubating module

**Comment by HarishBalachander** [comment]
Any timeline on when this feature will be released ?

**Comment by smirnp** [comment]
I faced a problem, than on actual codes the return of "snapshots compiler" is not fully correct. So decicided to use not the latest codes. And I down't know, them will have an ability to modify it up to the actual codes.

**Comment by bartoszwalacik** [comment]
of course, before we can merge, you need to catch up with our master

**Comment by bartoszwalacik** [comment]
we have decided to create SnapshotsCompiler feature in a different way

**Pull request closed without merging**